### ISSUE TITLE
Should the default for non-nullable reference types really be null?

### ISSUE THREAD

**Issue by AustinBryan** [dotnet/csharplang#1610]
Doing:

```csharp
string s = default;
```

Gives warning that it should be `string?` instead. I would've thought `default` to be `""`. 

I get that you guys don't want breaking changes, and if they were expect `default` to be `null` before, then it should still be, but at the same time, isn't this an opt-in feature?

It would be nice if `default` for non-nullable reference types automatically called the default constructor. I get that you guys are considering doing `Button button = new` which would do that, but that's not the same what I'd expect from `default`. 

I'd expect nullable types to default to `null` and non-nullable to be `new`, `0`, `(0, 0)`, etc.

**Comment by ethanae** [comment]
Remember `System.String` is a reference type and is certainly nullable so its `default` is null, hence the compiler warning.

**Comment by AustinBryan** [comment]
@ethanae I get that. But `System.Int32` is a value type, which is not nullable, yet the `default` for `int` is `0` and `default`  for `int?` is `null` for that to work and not the other way around is a bit inconsistent. 

Shouldn't the compiler see the lack of `?` and give a `default` that's non `null`, such as `""` or `new()`?

**Comment by ufcpp** [comment]
Not only this but also many inconsistency, for instance:

```cs
int? x = null;
var s = x.ToString(); // returns empty string, doesn't throw NRE
```

**Comment by ethanae** [comment]
@AustinBryan I understand - however, the ability for value types to be null is because of `Nullable<T>`.  So it is easier to distinguish what the default for it should be. All references, by default, have this behaviour and so naturally `?` actually has no effect because it is implied. 

I agree it is a bit inconsistent between `string` and a user defined reference type. `MyClass` cannot be defined as nullable explicitly as `MyClass? myClass;` but the syntax works with `string`. This may be due to the thinking that `string` actually acts like a value type under the hood and so in a sense it remains consistent with `string`'s value type behaviour.

**Comment by zippec** [comment]
I think that `default` should always only zero-out the memory slot for the type. It should be as cheap as possible and it definitely should not call any arbitrarily large piece of code (e.g. constructor) under the hood.

**Comment by DavidArno** [comment]
@zippec, that is what happens currently. `default` is always 0-filled. That's why `string s = default;` has to give a warning. There is no current way of it being anything else without huge performance impacts.

**Comment by DavidArno** [comment]
@AustinBryan,

Out of curiosity, if I have the type,
```cs
class SomeType
{
    public object A {get;}
    public string B {get;}
    public Regex C {get;}

    public SomeType(object a, string b, Regex c) => (A, B, C) = (a, b, c);
}
```
What would you expect `SomeType s = default;` to do here, other than `default` to be `null` and a warning to be emitted?

**Comment by bondsbw** [comment]
One possibility, `default(T)` could be lowered to `new T()`, which is only possible if `T : new()` (otherwise error).

**Comment by gafter** [comment]
> One possibility, `default(T)` could be lowered to `new T()`, which is only possible if `T : new()` (otherwise error).

That isn't a plausible language change, because `default(T)` is already legal and that is not what it means.

**Comment by bondsbw** [comment]
@gafter  I'm just talking about when `T` is a non-nullable reference type.

NRT off:
```c#
string s = default;  // s == null
```

NRT on:
```c#
string? s = default; // s == null
string t = default;  // t == ""
```

That may be more of a breaking change when switching on NRT than the LDM wants to consider.  But I thought I'd throw it out there.

**Comment by MgSam** [comment]
`default` already has a meaning. There is zero chance that meaning will ever change- as it would be a huge breaking change.

Nullable reference types do not affect semantics.

**Comment by AustinBryan** [comment]
@DavidArno  I thought about that, and my thinking is that it's okay for the objects *inside* the non-nullable type to be `null` when using `default`, but the overall object itself won't be. That would be the cheapest alternative (next to the whole instance being `null`). 

Other than that, it would call the default constructor all the way down, which could be a huge performance hit so that would be a bad idea. 

Here's my idea:

```csharp
class SomeType
{
    public object A { get; }
    public string B { get; }
    public Regex C { get; }

    public SomeType(object a, string b, Regex c) => (A, B, C) = (a, b, c);
}

...

SomeType someType = default;    // Calls SomeType() { } even though the user didn't define one
someType.B = default;

Console.WriteLine(someType == null);    // false
Console.WriteLine(someType.A == null);  // true
Console.WriteLine(someType.B == null);  // false
```

Basically, `default` is doing the least amount of work possible to make sure, at the very least `someType` is not null, but it may contain `null`. 

Maybe there could be some `defaultAll` or `noNull` that would do the behavior of making sure the object and nothing in it is `null`.

**Comment by AustinBryan** [comment]
@MgSam But it's an opt-in change. If not, then what about a new word that would automatically call a default constructor, a cheap one that only creates the class and not the objects in it so that the object itself won't be `null` if it isn't nullable, like `full` or `notNull` or something?

**Comment by yaakov-h** [comment]
That sounds awfully dangerous, you would bypass any constructor validation and invalidate any invariants.

For structs, empty/default is a well-known and understood case. For classes this doesn't exist, and changing this would do the exact opposite of what NRT is trying to achieve - better compile-time safety.

**Comment by jnm2** [comment]
And if you want to do this awfully dangerous, invariant-invalidating hack:
https://docs.microsoft.com/dotnet/api/system.runtime.serialization.formatterservices.getuninitializedobject
It's just as much a foot gun as the rest of reflection.

**Comment by asyncawaitmvvm** [comment]
> But it's an opt-in change.

Good. Because otherwise you will end up with a fork.

**Comment by paulomorgado** [comment]
@yaakov-h, up until structs get parameterless constructors when `default(S) != new S()`.

**Comment by yaakov-h** [comment]
@paulomorgado IIRC the LDM said that’s off the cards. I think it’s possible in IL, but since it’s not in a mainstream CLR language you can’t derive any widespread meaning from it.

**Comment by Joe4evr** [comment]
You can already write parameterless contructors for structs in IL since the dawn of the CLR and the languages respect it just fine. In that sense, `default(S)` and `new S()` have never been guaranteed to behave exactly the same.

**Comment by DavidArno** [comment]
Oh good, do we get to have this discussion again? Whilst `default(S)` and `new S()` have never been guaranteed to behave exactly the same, there is likely a *lot* of code out there that reasonably assumes they are the same as no mainstream .NET language supports defining them differently. 

So one should play with this feature in IL with extreme "here be dragons" caution and only make them different if really necessary.

**Comment by ufcpp** [comment]
No-arg struct constructor has already been championed, 8.0 candidate.
https://github.com/dotnet/csharplang/issues/99

**Comment by paulomorgado** [comment]
I know @Joe4evr. This is one of those that does not require CLR changes. I was just trying to not complicate.

**Comment by DavidArno** [comment]
@ufcpp, I thought there was some meeting notes recently that said they'd dropped the idea. Maybe I imagined it though.

**Comment by ufcpp** [comment]
@DavidArno 
Is it a recent note? I thought

- once no-arg ctor was implemented in C# 6.0
- it could not be released because of a **bug** that some classes treat `new S()` as `default(S)`
- it is still an 8.0 candidate and will be re-implemented after fixing the bug

**Comment by HaloFour** [comment]
@ufcpp 

> it could not be released because of a bug that some classes treat `new S()` as `default(S)`

It was a bug with Activator.CreateInstance:  https://github.com/dotnet/coreclr/pull/10778

**Comment by DavidArno** [comment]
I thought there'd been a more recent change of mind than that. But since it's still a v8 candidate, I must have imagined that.

**Comment by AustinBryan** [comment]
What if we could define what we wanted to be the default? Something like:

```csharp
class Foo
{
    public Foo() { }
    default => new Foo();
}
class Bar
{
    public Bar() { }
    public Bar(int x) { }

    default => new Bar(-1);
}
```

If the type is non-nullable `Bar`, it will use the custom-default of `new Bar(-1)`, but if it is `Bar?` it will use `null` instead. It will also use `null` if no `default` is specified.

**Comment by HaloFour** [comment]
@AustinBryan 

`default` is explicitly defined as "zeroing-out" because that's the only option that makes sense when initializing locals or arrays, particularly when generics are concerned.

```cs
public T Baz<T>() => default(T);

var bar = Baz<Bar>(); // neither the compiler nor the CLR will be able to call `Bar::default`
```

**Comment by DavidArno** [comment]
@HaloFour, I completely disagree that “zeroing out” is the only thing that makes sense. It’s simple and fast. That is all.

**Comment by HaloFour** [comment]
@DavidArno 

> It’s simple and fast. That is all.

And that is precisely why it's the only option that makes sense.  With generics there really isn't other options, not without changing the CLR to specialize for every generic type argument and attempt to bind to a `default` method at runtime.  This horse has been well tenderized.

**Comment by CyrusNajmabadi** [comment]
I agree with @HaloFour .  Note: if you do want a solution here, you might be able to get one with the Shapes/HKTs proposals.  There you could have some sort of SDefaultable shape.  And you can then define, per type, however you want a 'domain-default' value to be created for it.  Then, you can write your generic algorithms that work on your types, but create a 'domain default' you think is appropriate, as opposed to creating just a CLR-default value.

**Comment by DavidArno** [comment]
@CyrusNajmabadi,

I’m confused by your response. On the one hand you say you agree with @HaloFour that there’s only one sensible option. Then in the next breath say that HKT’s/shapes would provide another option.

Is there only one option or not?

**Comment by CyrusNajmabadi** [comment]
> I’m confused by your response

I'm confused by your confusion:

>On the one hand you say you agree with @HaloFour that there’s only one sensible option.

What was in reference to what ```default(T)``` means. 

> Then in the next breath say that HKT’s/shapes would provide another option.

That's in reference to how a developer (like yourself), could have an idea of a "domain specific 'default'" that htey might want to build an API around.  i.e. you could define libraries that allowed people  to inject what they wanted as some sort big-D 'Default' value.  Shapes would help facilitate that.  That sort of big-D 'Default' is not the same as the little-d ```default(T)``` which i think should always have CLR semantics.

**Comment by DavidArno** [comment]
@CyrusNajmabadi,

Thanks for the clarification. I'm no longer confused.

> That's in reference to how a developer (like yourself), could have an idea of a "domain specific 'default'" that htey might want to build an API around. i.e. you could define libraries that allowed people to inject what they wanted as some sort big-D 'Default' value. Shapes would help facilitate that. 

We already have that: `new T(...)`. This thread was about the behaviour of `default(T)` and how zero-fill increasingly makes less sense, but - as @HaloFour says - we are stuck with it as changing that in the CLR would be too big a job.

**Comment by YairHalberstadt** [comment]
I'm going to close this, as what is suggested here would be a hugely breaking change. You suggest linking it to nullable reference types, but they have already been released by now without this change, so making this change now would be breaking.

If you would like to discuss this further, let me know and I'll move it to a discussion.
